ql/src/java/org/apache/hadoop/hive/ql/parse/rewrite/CopyOnWriteUpdateRewriter.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.parse.rewrite;
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.api.FieldSchema;
+import org.apache.hadoop.hive.ql.Context;
+import org.apache.hadoop.hive.ql.metadata.HiveUtils;
+import org.apache.hadoop.hive.ql.parse.ASTNode;
+import org.apache.hadoop.hive.ql.parse.CalcitePlanner;
+import org.apache.hadoop.hive.ql.parse.HiveParser;
+import org.apache.hadoop.hive.ql.parse.ParseUtils;
+import org.apache.hadoop.hive.ql.parse.SemanticException;
+import org.apache.hadoop.hive.ql.parse.rewrite.sql.COWWithClauseBuilder;
+import org.apache.hadoop.hive.ql.parse.rewrite.sql.MultiInsertSqlBuilder;
+import org.apache.hadoop.hive.ql.parse.rewrite.sql.SetClausePatcher;
+import org.apache.hadoop.hive.ql.parse.rewrite.sql.SqlBuilderFactory;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class CopyOnWriteUpdateRewriter implements Rewriter<UpdateStatement> {
+
+  private final HiveConf conf;
+  private final SqlBuilderFactory sqlBuilderFactory;
+  private final COWWithClauseBuilder cowWithClauseBuilder;
+  private final SetClausePatcher setClausePatcher;
+
+
+  public CopyOnWriteUpdateRewriter(HiveConf conf, SqlBuilderFactory sqlBuilderFactory,
+                                   COWWithClauseBuilder cowWithClauseBuilder, SetClausePatcher setClausePatcher) {
+    this.conf = conf;
+    this.sqlBuilderFactory = sqlBuilderFactory;
+    this.cowWithClauseBuilder = cowWithClauseBuilder;
+    this.setClausePatcher = setClausePatcher;
+  }
+
+  @Override
+  public ParseUtils.ReparseResult rewrite(Context context, UpdateStatement updateBlock)
+      throws SemanticException {
+
+    String whereClause = context.getTokenRewriteStream().toString(
+        updateBlock.getWhereTree().getChild(0).getTokenStartIndex(),
```

Comment:
same reg local var

Reply:
Currently Update with empty where clause is not supported. We need to implement it in a follow-up patch.

Reply:
👍